第一章 概述/lidar_localization/third_part/GeographicLib-1.52/man/MagneticField.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.42)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MAGNETICFIELD 1"
.TH MAGNETICFIELD 1 "2021-06-21" "GeographicLib 1.52" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
MagneticField \-\- compute the earth's magnetic field
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBMagneticField\fR [ \fB\-n\fR \fIname\fR ] [ \fB\-d\fR \fIdir\fR ]
[ \fB\-N\fR \fINmax\fR ] [ \fB\-M\fR \fIMmax\fR ]
[ \fB\-t\fR \fItime\fR | \fB\-c\fR \fItime\fR \fIlat\fR \fIh\fR ]
[ \fB\-r\fR ] [ \fB\-w\fR ] [ \fB\-T\fR \fItguard\fR ] [ \fB\-H\fR \fIhguard\fR ] [ \fB\-p\fR \fIprec\fR ]
[ \fB\-v\fR ]
[ \fB\-\-comment\-delimiter\fR \fIcommentdelim\fR ]
[ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
[ \fB\-\-input\-file\fR \fIinfile\fR | \fB\-\-input\-string\fR \fIinstring\fR ]
[ \fB\-\-line\-separator\fR \fIlinesep\fR ]
[ \fB\-\-output\-file\fR \fIoutfile\fR ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBMagneticField\fR reads in times and positions on standard input and
prints out the geomagnetic field on standard output and, optionally, its
rate of change.
.PP
The input line is of the form \fItime\fR \fIlat\fR \fIlon\fR \fIh\fR. \fItime\fR is a
date of the form 2012\-07\-03, a fractional year such as 2012.5, or the
string \*(L"now\*(R".  \fIlat\fR and \fIlon\fR are the latitude and longitude
expressed as decimal degrees or degrees, minutes, and seconds; for
details on the allowed formats for latitude and longitude, see the
\&\f(CW\*(C`GEOGRAPHIC COORDINATES\*(C'\fR section of \fBGeoConvert\fR\|(1).  \fIh\fR is the height
above the ellipsoid in meters; this is optional and defaults to zero.
Alternatively, \fItime\fR can be given on the command line as the argument
to the \fB\-t\fR option, in which case it should not be included on the
input lines.  Finally, the magnetic field can be computed at various
points on a circle of latitude (constant \fItime\fR, \fIlat\fR, and \fIh\fR) via
the \fB\-c\fR option; in this case only the longitude should be given on the
input lines.
.PP
The output consists of the following 7 items:
.PP
.Vb 9
\&  the declination (the direction of the horizontal component of
\&    the magnetic field measured clockwise from north) in degrees,
\&  the inclination (the direction of the magnetic field measured
\&    down from the horizontal) in degrees,
\&  the horizontal field in nanotesla (nT),
\&  the north component of the field in nT,
\&  the east component of the field in nT,
\&  the vertical component of the field in nT (down is positive),
\&  the total field in nT.
.Ve
.PP
If the \fB\-r\fR option is given, a second line is printed giving the rates
of change of these quantities in degrees/yr and nT/yr.
.PP
The \s-1WGS84\s0 ellipsoid is used, \fIa\fR = 6378137 m, \fIf\fR = 1/298.257223563.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-n\fR \fIname\fR" 4
.IX Item "-n name"
use magnetic field model \fIname\fR instead of the default \f(CW\*(C`wmm2020\*(C'\fR.  See
\&\*(L"\s-1MODELS\*(R"\s0.
.IP "\fB\-d\fR \fIdir\fR" 4
.IX Item "-d dir"
read magnetic models from \fIdir\fR instead of the default.  See
\&\*(L"\s-1MODELS\*(R"\s0.
.IP "\fB\-N\fR \fINmax\fR" 4
.IX Item "-N Nmax"
limit the degree of the model to \fINmax\fR.
.IP "\fB\-M\fR \fIMmax\fR" 4
.IX Item "-M Mmax"
limit the order of the model to \fIMmax\fR.
.IP "\fB\-t\fR \fItime\fR" 4
.IX Item "-t time"
evaluate the field at \fItime\fR instead of reading the time from the input
lines.
.IP "\fB\-c\fR \fItime\fR \fIlat\fR \fIh\fR" 4
.IX Item "-c time lat h"
evaluate the field on a circle of latitude given by \fItime\fR, \fIlat\fR,
\&\fIh\fR instead of reading these quantities from the input lines.  In this
case, \fBMagneticField\fR can calculate the field considerably more
quickly.
.IP "\fB\-r\fR" 4
.IX Item "-r"
toggle whether to report the rates of change of the field.
.IP "\fB\-w\fR" 4
.IX Item "-w"
toggle the longitude first flag (it starts off); if the flag is on, then
on input and output, longitude precedes latitude (except that, on input,
this can be overridden by a hemisphere designator, \fIN\fR, \fIS\fR, \fIE\fR,
\&\fIW\fR).
.IP "\fB\-T\fR \fItguard\fR" 4
.IX Item "-T tguard"
signal an error if \fItime\fR lies \fItguard\fR years (default 50 yr) beyond
the range for the model.
.IP "\fB\-H\fR \fIhguard\fR" 4
.IX Item "-H hguard"
signal an error if \fIh\fR lies \fIhguard\fR meters (default 500000 m) beyond
the range for the model.
.IP "\fB\-p\fR \fIprec\fR" 4
.IX Item "-p prec"
set the output precision to \fIprec\fR (default 1).  Fields are printed
with precision with \fIprec\fR decimal places; angles use \fIprec\fR + 1
places.
.IP "\fB\-v\fR" 4
.IX Item "-v"
print information about the magnetic model on standard error before
processing the input.
.IP "\fB\-\-comment\-delimiter\fR \fIcommentdelim\fR" 4
.IX Item "--comment-delimiter commentdelim"
set the comment delimiter to \fIcommentdelim\fR (e.g., \*(L"#\*(R" or \*(L"//\*(R").  If
set, the input lines will be scanned for this delimiter and, if found,
the delimiter and the rest of the line will be removed prior to
processing and subsequently appended to the output line (separated by a
space).
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
print version and exit.
.IP "\fB\-h\fR" 4
.IX Item "-h"
print usage, the default magnetic path and name, and exit.
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
print full documentation and exit.
.IP "\fB\-\-input\-file\fR \fIinfile\fR" 4
.IX Item "--input-file infile"
read input from the file \fIinfile\fR instead of from standard input; a file
name of \*(L"\-\*(R" stands for standard input.
.IP "\fB\-\-input\-string\fR \fIinstring\fR" 4
.IX Item "--input-string instring"
read input from the string \fIinstring\fR instead of from standard input.
All occurrences of the line separator character (default is a semicolon)
in \fIinstring\fR are converted to newlines before the reading begins.
.IP "\fB\-\-line\-separator\fR \fIlinesep\fR" 4
.IX Item "--line-separator linesep"
set the line separator character to \fIlinesep\fR.  By default this is a
semicolon.
.IP "\fB\-\-output\-file\fR \fIoutfile\fR" 4
.IX Item "--output-file outfile"
write output to the file \fIoutfile\fR instead of to standard output; a
file name of \*(L"\-\*(R" stands for standard output.
.SH "MODELS"
.IX Header "MODELS"
\&\fBMagneticField\fR computes the geomagnetic field using one of the
following models
.PP
.Vb 10
\&    wmm2010, the World Magnetic Model 2010, which approximates the
\&      main magnetic field for the period 2010\-2015.  See
\&      https://ngdc.noaa.gov/geomag/WMM/DoDWMM.shtml
\&    wmm2015v2, the World Magnetic Model 2015, which approximates the
\&      main magnetic field for the period 2015\-2020.  See
\&      https://ngdc.noaa.gov/geomag/WMM/DoDWMM.shtml
\&    wmm2015, a deprecated version of wmm2015v2
\&    wmm2020, the World Magnetic Model 2020, which approximates the
\&      main magnetic field for the period 2020\-2025.  See
\&      https://ngdc.noaa.gov/geomag/WMM/DoDWMM.shtml
\&    igrf11, the International Geomagnetic Reference Field (11th
\&      generation), which approximates the main magnetic field for
\&      the period 1900\-2015.  See
\&      https://ngdc.noaa.gov/IAGA/vmod/igrf.html
\&    igrf12, the International Geomagnetic Reference Field (12th
\&      generation), which approximates the main magnetic field for
\&      the period 1900\-2020.  See
\&      https://ngdc.noaa.gov/IAGA/vmod/igrf.html
\&    igrf13, the International Geomagnetic Reference Field (13th
\&      generation), which approximates the main magnetic field for
\&      the period 1900\-2025.  See
\&      https://ngdc.noaa.gov/IAGA/vmod/igrf.html
\&    emm2010, the Enhanced Magnetic Model 2010, which approximates
\&      the main and crustal magnetic fields for the period 2010\-2015.
\&      See https://ngdc.noaa.gov/geomag/EMM/index.html
\&    emm2015, the Enhanced Magnetic Model 2015, which approximates
\&      the main and crustal magnetic fields for the period 2000\-2020.
\&      See https://ngdc.noaa.gov/geomag/EMM/index.html
\&    emm2017, the Enhanced Magnetic Model 2017, which approximates
\&      the main and crustal magnetic fields for the period 2000\-2022.
\&      See https://ngdc.noaa.gov/geomag/EMM/index.html
.Ve
.PP
These models approximate the magnetic field due to the earth's core and
(in the case of emm20xx) its crust.  They neglect magnetic fields due to
the ionosphere, the magnetosphere, nearby magnetized materials,
electrical machinery, etc.
.PP
By default, the \f(CW\*(C`wmm2020\*(C'\fR magnetic model is used.  This may changed by
setting the environment variable \f(CW\*(C`GEOGRAPHICLIB_MAGNETIC_NAME\*(C'\fR or with
the \fB\-n\fR option.
.PP
The magnetic models will be loaded from a directory specified at compile
time.  This may changed by setting the environment variables
\&\f(CW\*(C`GEOGRAPHICLIB_MAGNETIC_PATH\*(C'\fR or \f(CW\*(C`GEOGRAPHICLIB_DATA\*(C'\fR, or with the
\&\fB\-d\fR option.  The \fB\-h\fR option prints the default magnetic path and
name.  Use the \fB\-v\fR option to ascertain the full path name of the data
file.
.PP
Instructions for downloading and installing magnetic models are
available at
<https://geographiclib.sourceforge.io/html/magnetic.html#magneticinst>.
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
.IP "\fB\s-1GEOGRAPHICLIB_MAGNETIC_NAME\s0\fR" 4
.IX Item "GEOGRAPHICLIB_MAGNETIC_NAME"
Override the compile-time default magnetic name of \f(CW\*(C`wmm2020\*(C'\fR.  The
\&\fB\-h\fR option reports the value of \fB\s-1GEOGRAPHICLIB_MAGNETIC_NAME\s0\fR, if
defined, otherwise it reports the compile-time value.  If the \fB\-n\fR
\&\fIname\fR option is used, then \fIname\fR takes precedence.
.IP "\fB\s-1GEOGRAPHICLIB_MAGNETIC_PATH\s0\fR" 4
.IX Item "GEOGRAPHICLIB_MAGNETIC_PATH"
Override the compile-time default magnetic path.  This is typically
\&\f(CW\*(C`/usr/local/share/GeographicLib/magnetic\*(C'\fR on Unix-like systems and
\&\f(CW\*(C`C:/ProgramData/GeographicLib/magnetic\*(C'\fR on Windows systems.  The \fB\-h\fR
option reports the value of \fB\s-1GEOGRAPHICLIB_MAGNETIC_PATH\s0\fR, if defined,
otherwise it reports the compile-time value.  If the \fB\-d\fR \fIdir\fR option
is used, then \fIdir\fR takes precedence.
.IP "\fB\s-1GEOGRAPHICLIB_DATA\s0\fR" 4
.IX Item "GEOGRAPHICLIB_DATA"
Another way of overriding the compile-time default magnetic path.  If it
is set (and if \fB\s-1GEOGRAPHICLIB_MAGNETIC_PATH\s0\fR is not set), then
$\fB\s-1GEOGRAPHICLIB_DATA\s0\fR/magnetic is used.
.SH "ERRORS"
.IX Header "ERRORS"
An illegal line of input will print an error message to standard output
beginning with \f(CW\*(C`ERROR:\*(C'\fR and causes \fBMagneticField\fR to return an exit
code of 1.  However, an error does not cause \fBMagneticField\fR to
terminate; following lines will be converted.  If \fItime\fR or \fIh\fR are
outside the recommended ranges for the model (but inside the ranges
increase by \fItguard\fR and \fIhguard\fR), a warning is printed on standard
error and the field (which may be inaccurate) is returned in the normal
way.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
The magnetic field from \s-1WMM2020\s0 in Timbuktu on 2020\-01\-01
.PP
.Vb 3
\&    echo 2020\-01\-01 16:46:33N 3:00:34W 300 | MagneticField \-r
\&    => \-1.60 12.00 33973.5 33960.3 \-948.1 7223.0 34732.8
\&       0.13 \-0.02 21.8 23.9 77.9 \-8.4 19.5
.Ve
.PP
The first two numbers returned are the declination and inclination of
the field.  The second line gives the annual change.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBGeoConvert\fR\|(1), \fBgeographiclib\-get\-magnetic\fR\|(8).
.SH "AUTHOR"
.IX Header "AUTHOR"
\&\fBMagneticField\fR was written by Charles Karney.
.SH "HISTORY"
.IX Header "HISTORY"
\&\fBMagneticField\fR was added to GeographicLib,
<https://geographiclib.sourceforge.io>, in version 1.15.


第一章 概述/lidar_localization/third_part/GeographicLib-1.52/tools/MagneticField.cpp
/**
 * \file MagneticField.cpp
 * \brief Command line utility for evaluating magnetic fields
 *
 * Copyright (c) Charles Karney (2011-2020) <charles@karney.com> and licensed
 * under the MIT/X11 License.  For more information, see
 * https://geographiclib.sourceforge.io/
 *
 * See the <a href="MagneticField.1.html">man page</a> for usage information.
 **********************************************************************/

#include <iostream>
#include <string>
#include <sstream>
#include <fstream>
#include <GeographicLib/MagneticModel.hpp>
#include <GeographicLib/MagneticCircle.hpp>
#include <GeographicLib/DMS.hpp>
#include <GeographicLib/Utility.hpp>

#if defined(_MSC_VER)
// Squelch warnings about constant conditional expressions
#  pragma warning (disable: 4127)
#endif

#include "MagneticField.usage"

int main(int argc, const char* const argv[]) {
  try {
    using namespace GeographicLib;
    typedef Math::real real;
    Utility::set_digits();
    bool verbose = false, longfirst = false;
    std::string dir;
    std::string model = MagneticModel::DefaultMagneticName();
    std::string istring, ifile, ofile, cdelim;
    char lsep = ';';
    real time = 0, lat = 0, h = 0;
    bool timeset = false, circle = false, rate = false;
    real hguard = 500000, tguard = 50;
    int prec = 1, Nmax = -1, Mmax = -1;

    for (int m = 1; m < argc; ++m) {
      std::string arg(argv[m]);
      if (arg == "-n") {
        if (++m == argc) return usage(1, true);
        model = argv[m];
      } else if (arg == "-d") {
        if (++m == argc) return usage(1, true);
        dir = argv[m];
      } else if (arg == "-N") {
        if (++m == argc) return usage(1, true);
        try {
          Nmax = Utility::val<int>(std::string(argv[m]));
          if (Nmax < 0) {
            std::cerr << "Maximum degree " << argv[m] << " is negative\n";
            return 1;
          }
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "-M") {
        if (++m == argc) return usage(1, true);
        try {
          Mmax = Utility::val<int>(std::string(argv[m]));
          if (Mmax < 0) {
            std::cerr << "Maximum order " << argv[m] << " is negative\n";
            return 1;
          }
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "-t") {
        if (++m == argc) return usage(1, true);
        try {
          time = Utility::fractionalyear<real>(std::string(argv[m]));
          timeset = true;
          circle = false;
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding argument of " << arg << ": "
                    << e.what() << "\n";
          return 1;
        }
      } else if (arg == "-c") {
        if (m + 3 >= argc) return usage(1, true);
        try {
          using std::abs;
          time = Utility::fractionalyear<real>(std::string(argv[++m]));
          DMS::flag ind;
          lat = DMS::Decode(std::string(argv[++m]), ind);
          if (ind == DMS::LONGITUDE)
            throw GeographicErr("Bad hemisphere letter on latitude");
          if (!(abs(lat) <= 90))
            throw GeographicErr("Latitude not in [-90d, 90d]");
          h = Utility::val<real>(std::string(argv[++m]));
          timeset = false;
          circle = true;
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding argument of " << arg << ": "
                    << e.what() << "\n";
          return 1;
        }
      } else if (arg == "-r")
        rate = !rate;
      else if (arg == "-w")
        longfirst = !longfirst;
      else if (arg == "-p") {
        if (++m == argc) return usage(1, true);
        try {
          prec = Utility::val<int>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "-T") {
        if (++m == argc) return usage(1, true);
        try {
          tguard = Utility::val<real>(std::string(argv[m]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding argument of " << arg << ": "
                    << e.what() << "\n";
          return 1;
        }
      } else if (arg == "-H") {
        if (++m == argc) return usage(1, true);
        try {
          hguard = Utility::val<real>(std::string(argv[m]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding argument of " << arg << ": "
                    << e.what() << "\n";
          return 1;
        }
      } else if (arg == "-v")
        verbose = true;
      else if (arg == "--input-string") {
        if (++m == argc) return usage(1, true);
        istring = argv[m];
      } else if (arg == "--input-file") {
        if (++m == argc) return usage(1, true);
        ifile = argv[m];
      } else if (arg == "--output-file") {
        if (++m == argc) return usage(1, true);
        ofile = argv[m];
      } else if (arg == "--line-separator") {
        if (++m == argc) return usage(1, true);
        if (std::string(argv[m]).size() != 1) {
          std::cerr << "Line separator must be a single character\n";
          return 1;
        }
        lsep = argv[m][0];
      } else if (arg == "--comment-delimiter") {
        if (++m == argc) return usage(1, true);
        cdelim = argv[m];
      } else if (arg == "--version") {
        std::cout << argv[0] << ": GeographicLib version "
                  << GEOGRAPHICLIB_VERSION_STRING << "\n";
        return 0;
      } else {
        int retval = usage(!(arg == "-h" || arg == "--help"), arg != "--help");
        if (arg == "-h")
          std::cout<< "\nDefault magnetic path = \""
                   << MagneticModel::DefaultMagneticPath()
                   << "\"\nDefault magnetic name = \""
                   << MagneticModel::DefaultMagneticName()
                   << "\"\n";
        return retval;
      }
    }

    if (!ifile.empty() && !istring.empty()) {
      std::cerr << "Cannot specify --input-string and --input-file together\n";
      return 1;
    }
    if (ifile == "-") ifile.clear();
    std::ifstream infile;
    std::istringstream instring;
    if (!ifile.empty()) {
      infile.open(ifile.c_str());
      if (!infile.is_open()) {
        std::cerr << "Cannot open " << ifile << " for reading\n";
        return 1;
      }
    } else if (!istring.empty()) {
      std::string::size_type m = 0;
      while (true) {
        m = istring.find(lsep, m);
        if (m == std::string::npos)
          break;
        istring[m] = '\n';
      }
      instring.str(istring);
    }
    std::istream* input = !ifile.empty() ? &infile :
      (!istring.empty() ? &instring : &std::cin);

    std::ofstream outfile;
    if (ofile == "-") ofile.clear();
    if (!ofile.empty()) {
      outfile.open(ofile.c_str());
      if (!outfile.is_open()) {
        std::cerr << "Cannot open " << ofile << " for writing\n";
        return 1;
      }
    }
    std::ostream* output = !ofile.empty() ? &outfile : &std::cout;

    tguard = std::max(real(0), tguard);
    hguard = std::max(real(0), hguard);
    prec = std::min(10 + Math::extra_digits(), std::max(0, prec));
    int retval = 0;
    try {
      using std::isfinite;
      const MagneticModel m(model, dir, Geocentric::WGS84(), Nmax, Mmax);
      if ((timeset || circle)
          && (!isfinite(time) ||
              time < m.MinTime() - tguard ||
              time > m.MaxTime() + tguard))
        throw GeographicErr("Time " + Utility::str(time) +
                            " too far outside allowed range [" +
                            Utility::str(m.MinTime()) + "," +
                            Utility::str(m.MaxTime()) + "]");
      if (circle
          && (!isfinite(h) ||
              h < m.MinHeight() - hguard ||
              h > m.MaxHeight() + hguard))
        throw GeographicErr("Height " + Utility::str(h/1000) +
                            "km too far outside allowed range [" +
                            Utility::str(m.MinHeight()/1000) + "km," +
                            Utility::str(m.MaxHeight()/1000) + "km]");
      if (verbose) {
        std::cerr << "Magnetic file: " << m.MagneticFile()      << "\n"
                  << "Name: "          << m.MagneticModelName() << "\n"
                  << "Description: "   << m.Description()       << "\n"
                  << "Date & Time: "   << m.DateTime()          << "\n"
                  << "Time range: ["
                  << m.MinTime() << ","
                  << m.MaxTime() << "]\n"
                  << "Height range: ["
                  << m.MinHeight()/1000 << "km,"
                  << m.MaxHeight()/1000 << "km]\n";
      }
      if ((timeset || circle) && (time < m.MinTime() || time > m.MaxTime()))
        std::cerr << "WARNING: Time " << time
                  << " outside allowed range ["
                  << m.MinTime() << "," << m.MaxTime() << "]\n";
      if (circle && (h < m.MinHeight() || h > m.MaxHeight()))
        std::cerr << "WARNING: Height " << h/1000
                  << "km outside allowed range ["
                  << m.MinHeight()/1000 << "km,"
                  << m.MaxHeight()/1000 << "km]\n";
      const MagneticCircle c(circle ? m.Circle(time, lat, h) :
                             MagneticCircle());
      std::string s, eol, stra, strb;
      std::istringstream str;
      while (std::getline(*input, s)) {
        try {
          eol = "\n";
          if (!cdelim.empty()) {
            std::string::size_type n = s.find(cdelim);
            if (n != std::string::npos) {
              eol = " " + s.substr(n) + "\n";
              s = s.substr(0, n);
            }
          }
          str.clear(); str.str(s);
          if (!(timeset || circle)) {
            if (!(str >> stra))
              throw GeographicErr("Incomplete input: " + s);
            time = Utility::fractionalyear<real>(stra);
            if (time < m.MinTime() - tguard || time > m.MaxTime() + tguard)
              throw GeographicErr("Time " + Utility::str(time) +
                                  " too far outside allowed range [" +
                                  Utility::str(m.MinTime()) + "," +
                                  Utility::str(m.MaxTime()) +
                                  "]");
            if (time < m.MinTime() || time > m.MaxTime())
              std::cerr << "WARNING: Time " << time
                        << " outside allowed range ["
                        << m.MinTime() << "," << m.MaxTime() << "]\n";
          }
          real lon;
          if (circle) {
            if (!(str >> strb))
              throw GeographicErr("Incomplete input: " + s);
            DMS::flag ind;
            lon = DMS::Decode(strb, ind);
            if (ind == DMS::LATITUDE)
              throw GeographicErr("Bad hemisphere letter on " + strb);
          } else {
            if (!(str >> stra >> strb))
              throw GeographicErr("Incomplete input: " + s);
            DMS::DecodeLatLon(stra, strb, lat, lon, longfirst);
            h = 0;              // h is optional
            if (str >> h) {
              if (h < m.MinHeight() - hguard || h > m.MaxHeight() + hguard)
                throw GeographicErr("Height " + Utility::str(h/1000) +
                                    "km too far outside allowed range [" +
                                    Utility::str(m.MinHeight()/1000) + "km," +
                                    Utility::str(m.MaxHeight()/1000) + "km]");
              if (h < m.MinHeight() || h > m.MaxHeight())
                std::cerr << "WARNING: Height " << h/1000
                          << "km outside allowed range ["
                          << m.MinHeight()/1000 << "km,"
                          << m.MaxHeight()/1000 << "km]\n";
            }
            else
              str.clear();
          }
          if (str >> stra)
            throw GeographicErr("Extra junk in input: " + s);
          real bx, by, bz, bxt, byt, bzt;
          if (circle)
            c(lon, bx, by, bz, bxt, byt, bzt);
          else
            m(time, lat, lon, h, bx, by, bz, bxt, byt, bzt);
          real H, F, D, I, Ht, Ft, Dt, It;
          MagneticModel::FieldComponents(bx, by, bz, bxt, byt, bzt,
                                         H, F, D, I, Ht, Ft, Dt, It);

          *output << DMS::Encode(D, prec + 1, DMS::NUMBER) << " "
                  << DMS::Encode(I, prec + 1, DMS::NUMBER) << " "
                  << Utility::str(H, prec) << " "
                  << Utility::str(by, prec) << " "
                  << Utility::str(bx, prec) << " "
                  << Utility::str(-bz, prec) << " "
                  << Utility::str(F, prec) << eol;
          if (rate)
            *output << DMS::Encode(Dt, prec + 1, DMS::NUMBER) << " "
                    << DMS::Encode(It, prec + 1, DMS::NUMBER) << " "
                    << Utility::str(Ht, prec) << " "
                    << Utility::str(byt, prec) << " "
                    << Utility::str(bxt, prec) << " "
                    << Utility::str(-bzt, prec) << " "
                    << Utility::str(Ft, prec) << eol;
        }
        catch (const std::exception& e) {
          *output << "ERROR: " << e.what() << "\n";
          retval = 1;
        }
      }
    }
    catch (const std::exception& e) {
      std::cerr << "Error reading " << model << ": " << e.what() << "\n";
      retval = 1;
    }
    return retval;
  }
  catch (const std::exception& e) {
    std::cerr << "Caught exception: " << e.what() << "\n";
    return 1;
  }
  catch (...) {
    std::cerr << "Caught unknown exception\n";
    return 1;
  }
}
